### ISSUE TITLE
WARNING:tensorflow:Your input ran out of data; interrupting training always occuring after 80% of total epochs

### ISSUE THREAD

**Issue by wzhou789** [tensorflow/tensorflow#64476]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

2.15

### Custom code

Yes

### Python version

3.10.12

### Current behavior?

While I'm training a tensorflow model, the training keeps gets interrupted at the epoch after completing 80% of the total number of epochs I specified to train for. For example, it will stop training at epoch 81 if I had set epochs to 100, with the warning "WARNING:tensorflow:Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches". 
I calculated steps_per_epoch as len(train_size) // batch_size. I have also tried using (len(train_size)//batch_size)-1. Both of these continue to cause the error specified above. I have also tried to leave out steps_per_epoch, which causes the model to start training with the validation steps per epoch instead. 

### Standalone code to reproduce the issue

```shell
https://colab.research.google.com/drive/1zNuEKH-K9pip9UonoOZC_2a1T5KdHFt7?usp=sharing
```

### Relevant log output

```shell
Epoch 300: val_loss did not improve from 1.24497
149/149 [==============================] - 13s 87ms/step - loss: 0.3701 - val_loss: 1.9121
Epoch 301/376
149/149 [==============================] - ETA: 0s - loss: 0.3454
Epoch 301: val_loss did not improve from 1.24497
149/149 [==============================] - 13s 85ms/step - loss: 0.3454 - val_loss: 3.6483
Epoch 302/376
149/149 [==============================] - ETA: 0s - loss: 0.3932
Epoch 302: val_loss did not improve from 1.24497
149/149 [==============================] - 13s 85ms/step - loss: 0.3932 - val_loss: 1.7741
Epoch 303/376
122/149 [=======================>......] - ETA: 1s - loss: 0.3708WARNING:tensorflow:Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches (in this case, 56024 batches). You may need to use the repeat() function when building your dataset.
```

**Comment by sushreebarsa** [comment]
@wzhou789 While your calculation (len(train_size) // batch_size) seems correct, there might be a subtle off-by-one error. Could you try subtracting 1 from your calculated steps_per_epoch and let us know?
```
steps_per_epoch = (len(train_size) // batch_size) - 1

```
Thank you!

**Comment by wzhou789** [comment]
@sushreebarsa I have already tried steps_per_epoch = (len(train_size) // batch_size) - 1. It gives the same error, just slightly more into the epochs after the run with the calculation as (len(train_size)//batch_size).

**Comment by sushreebarsa** [comment]
@wzhou789 Could you please ensure that your data is shuffled before being fed to the model during training. Shuffling randomizes the order of data points within each batch and across epochs.
Thank you!
